### DISCUSSION TITLE
useMemo usage

### DISCUSSION THREAD

**Discussion by AjaxSolutions** [TanStack/query#1641] [Q&A]
There is a lot of interesting code in React Query.  This library is helpful and well done.

I'm curious why _useMemo_ is used in the [useHydrate](https://github.com/tannerlinsley/react-query/blob/master/src/hydration/react.tsx#L15) hook as opposed to _useEffect_.  What is the benefit?

_useMemo_ is typically used to memoize the return value of a function, but the _useMemo_ function is not returning any value.

`React.useMemo(() => {
	if (state) {
	  hydrate(queryClient, state, optionsRef.current)
	}
}, [queryClient, state])`

**Comment by TkDodo** [comment]
I'm not sure, but it really looks like we're executing a `side-effect` here (building the queryCache), which should probably _not_ be done during render, because render should be a pure function... 🤔 

/edit: to clarify: I think `useEffect` sounds right, but I'm not sure of the implications

**Comment by AjaxSolutions** [comment]
I think this should be changed to use _useEffect_, unless there is a very compelling reason to use _useMemo_.

The function used by _useMemo_ should be a [pure function](https://github.com/reactjs/reactjs.org/issues/3348), but in this case the function is not pure.

**Reply by TkDodo** [reply]
can you PR this and test if it still works? I have not used the hydration feature before... maybe this would result in one render-cycle where we have no cache yet (because effects run after render), triggering all the queries as if they were new (or so ...)

**Reply by AjaxSolutions** [reply]
Would it make more sense to use useLayoutEffect since [useEffect](https://github.com/tannerlinsley/react-query/pull/1643) didn't work?
